titanium-rdf-api-java/src/main/java/eu/neverblink/jelly/convert/titanium/internal/TitaniumNode.java

```diff
@@ -0,0 +1,74 @@
+package eu.neverblink.jelly.convert.titanium.internal;
+
+import com.apicatalog.rdf.api.RdfQuadConsumer;
+import eu.neverblink.jelly.core.InternalApi;
+
+/**
+ * Internal representations of RDF data inside the Titanium converter.
+ * <p>
+ * These are not intended to be used outside of the converter's code.
+ */
+@InternalApi
+public sealed interface TitaniumNode {
+    enum TitaniumNodeType {
+        IRI,
+        BLANK,
+        SIMPLE_LITERAL,
+        LANG_LITERAL,
+        DT_LITERAL,
+    }
+
+    TitaniumNodeType type();
+
+    default StringNode asString() {
+        return (StringNode) this;
+    }
+
+    default String asStringValue() {
+        return asString().value();
+    }
+
+    default SimpleLiteral asSimpleLiteral() {
+        return (SimpleLiteral) this;
+    }
+
+    default LangLiteral asLangLiteral() {
+        return (LangLiteral) this;
+    }
+
+    default DtLiteral asDtLiteral() {
+        return (DtLiteral) this;
+    }
+
+    sealed interface TitaniumLiteral extends TitaniumNode {}
+
+    // String used to represent IRIs and blank nodes
+    record StringNode(String value) implements TitaniumNode {
```

Comment:
I'll be grumpy about this one. We should not encapsulate the String in anything else, if not needed. I know that in Java we can't do algebraic datatypes, but then please just use Object and instance of checks.

Reply:
The fewer allocations, the better :)